### ISSUE TITLE
OSX hangs at "calling kinit .." when ansible_winrm_kinit_mode set to managed

### ISSUE THREAD

**Issue by jameszh** [ansible/ansible#25555]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

windows kerberos
##### ANSIBLE VERSION

```
ansible 2.3.1.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.11 (default, Jan 22 2016, 08:29:18) [GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
```

##### OS / ENVIRONMENT

MacOS 10.12.5
##### SUMMARY

when ansible_winrm_kinit_mode=managed
it's stuck with following:
creating Kerberos CC at /var/folders/5k/m9hvl9ld2wv8zs33vhlrhvw40000gn/T/tmpILJqaA
calling kinit for principal xxxx@DOMAMAIN.COM
but when when ansible_winrm_kinit_mode=manual, and kinit manually.
it runs successfully.

##### STEPS TO REPRODUCE

```
ansible_connection=winrm
ansible_user=user@DOMAIN.COM
ansible_password=xxx
ansible_winrm_kinit_mode=managed
ansible_port=5986
ansible_winrm_transport=kerberos
ansible_winrm_server_cert_validation=ignore

```
##### ACTUAL RESULTS

See below

```
it's stuck with following:
creating Kerberos CC at /var/folders/5k/m9hvl9ld2wv8zs33vhlrhvw40000gn/T/tmpILJqaA
calling kinit for principal xxxx@DOMAMAIN.COM
```

**Comment by jctanner** [comment]
cc @nitzmahone

**Comment by nitzmahone** [comment]
Curious if this is 100% repeatable or if it works sometimes? I've never run into this, but I only tested briefly on Heimdal Kerberos (which is what I assume you're using, since you're on OSX). Is there any chance you can run from source and try a little experiment for me?

**Comment by jameszh** [comment]
100% repeatable, yes let me know how to do your experiment.

Thanks!

--
James Zhang
Sent with Airmail

**Comment by nitzmahone** [comment]
OK, cool- are you running Ansible from source?

**Comment by nitzmahone** [comment]
Once you are, find the line `stdout, stderr = p.communicate(password + b'\n')` in lib/ansible/plugins/winrm.py (in the _kerb_auth function, around line 182), and just before that, stick in 
```
import time
time.sleep(2)
```

then try your test again (making sure you're using the newly-hacked from-source version of Ansible, of course). 

This isn't of course a real fix, but just to narrow down the cause. I suspect there may be something special about either Heimdal or OSX where the input buffer is getting cleared before prompting for the kinit password. On Linux with MIT Kerberos, this doesn't occur anywhere I've tried it (and I've tried injecting significant delays to make sure), so we don't actually inspect stdout for the password prompt before shoving the password into the stdin buffer.

Also, just out of curiosity- is your KDC local to the Mac (eg a local Windows VM) or somewhere else? Is there a perceptible delay between running `kinit someuser@domain.com` and the password prompt appearing?

Thanks...

**Comment by jameszh** [comment]
same behaviour, here is the setup.
Cloned the latest repo: ansible 2.4.0 (devel 9d23dcac29) last updated 2017/06/09 14:19:31 (GMT -400)

I insert the following lines in the winrm.py as instructed.

        import time
        time.sleep(2)
        print(password)
        stdout, stderr = p.communicate(password + b'\n')

I do see it the printout(editing the right file), the code is hanging on p.communicate call (have to ctrl-c to break)

Yes, all running from local to the Mac, kinit prompt for password is instant.

Thanks,

**Comment by jborean93** [comment]
This has been fixed with https://github.com/ansible/ansible/commit/92e52ef5154fc78758cbc3bbb3edac95d81f46e0 and will be available in 2.5. Note you will need the `pexpect` Python module to be installed for it to work.
